rust/kernel/str.rs

```diff
@@ -500,3 +500,77 @@ impl fmt::Write for Formatter {
         }
     }
 }
+
+/// An owned string that is guaranteed to have exactly one `NUL` byte, which is at the end.
+///
+/// Used for interoperability with kernel APIs that take C strings.
+///
+/// # Examples
+///
+/// ```
+/// # use kernel::prelude::*;
+/// use kernel::str::CString;
+///
+/// let s = CString::try_from_fmt(fmt!("{}{}{}", "abc", 10, 20)).unwrap();
+/// assert_eq!(s.as_bytes_with_nul(), "abc1020\0".as_bytes());
+///
+/// let tmp = "testing";
+/// let s = CString::try_from_fmt(fmt!("{tmp}{}", 123)).unwrap();
+/// assert_eq!(s.as_bytes_with_nul(), "testing123\0".as_bytes());
+/// ```
+pub struct CString {
+    buf: Vec<u8>,
+}
+
+impl CString {
+    /// Creates an instance of [`CString`] from the given formatted arguments.
+    pub fn try_from_fmt(args: fmt::Arguments<'_>) -> Result<Self, Error> {
+        // Calculate the size needed (formatted string plus `NUL` terminator). Since we're
+        // initialising the initial buffer to `null`, we know the insert position will hold the
+        // required length after we've written everything.
+        // SAFETY: The buffer is empty, so we guarantee its validity vacuously.
+        let mut f =
+            unsafe { RawFormatter::from_ptrs(core::ptr::null_mut(), core::ptr::null_mut()) };
+        f.write_fmt(args)?;
+        f.write_str("\0")?;
+        let size = f.pos() as _;
+
+        // Allocate a vector with the required number of bytes, and write to it.
+        let mut buf = Vec::try_with_capacity(size)?;
+        // SAFETY: The buffer stored in `buf` is at least of size `size` and is valid for writes.
+        let mut f = unsafe { Formatter::from_buffer(buf.as_mut_ptr(), size) };
+        f.write_fmt(args)?;
+        f.write_str("\0")?;
+
+        // SAFETY: `size` is exactly the capacity of `buf`, and all elements have been initialised
+        // by the formatter above.
+        unsafe { buf.set_len(size) };
```

Comment:
There is no guarantee that the second time formatting will result in a string as long as the previous time, right? A `Display` or `Debug` impl may alternate between writing no string and writing a long string.

Reply:
Indeed. I changed this to use `f.bytes_written()`.

If a longer string is needed on the second round, the function fails.